### PR TITLE
Adds support for machinery to be bigger than 1x1

### PR DESCRIPTION
**Pull request by DGamerL**
## What Does This PR Do
Adds support for machinery to be bigger than a 1x1
Also removes an unneeded variable on `/obj/structure/filler`

### How does this system work
Add the component to the machinery on `Initialize()`, and add a list like the one below as the second argument. This list is a representation for a coordinate map around the machinery, and can be any size (3x3, 5x5, 7x7) as long as it has an uneven length and is square.
```js
	list(
			list(0, 0, 0,   0, 0),
			list(0, 0, 0,   0, 0),
			list(0, 0, MACH_CENTER, 0, 0),
			list(0, 0, 0,   0, 0),
			list(0, 0, 0,   0, 0)
		)
```

If you want to make specific turfs dense, you just have to change the int to `1`, which will make a filler turf there. 
All these filler turfs are automatically cleared upon `Destroy` and don't need any special overrides to be cleared out.

## Why It's Good For The Game
This allows us to make bigger machinery in various special shapes with a lot of freedom

## Testing
No rubber duckies for testing sadly, you're gonna have to trust that I touched these machines
image

I also tested the gravity generator, just didn't take a picture

<hr>

### Declaration
- [x] I confirm that I either do not require [pre-approval](/CODE_OF_CONDUCT.md#types-of-changes-that-need-approval) for this PR, or I have obtained such approval and have included a screenshot to demonstrate this below.

<hr>

## Changelog
Big thanks to beaglegaming for Component'izing it when I didn't have motivation to do so!
NPFC

### CODE REVIEW

File: code/__DEFINES/misc_defines.dm [hunk 1]
```diff
@@ -104,6 +104,9 @@
 #define STAGE_FIVE 9
 #define STAGE_SIX 11 //From supermatter shard
 
+/// A define for the center of the coordinate map of big machinery
+#define SRC 0
```
Comment (Contrabang) [inline code/__DEFINES/misc_defines.dm:108]: This is way too general of a name for a define that goes across the whole codebase, I suggest something like this
```suggestion
#define BIG_MACH_CENTER 0
```

**Review by Contrabang — CHANGES_REQUESTED** [review]

File: code/__DEFINES/misc_defines.dm (hunk 1 above)
Comment (DGamerL) [reply code/__DEFINES/misc_defines.dm:108]: Very good point!

**Comment by MLGTASTICa** [comment]
>     * Moving it to `/obj/machinery`, which _technically_ is fine but it'd give every piece of machinery a nested list with 25 integers, which can clutter VV a decent amount
> 

Why not just initialize the list only when needed? as in all the big machines just having a set filler proc() in their initialize, this way you avoid both unnecesarry subtypes and provide the functionality for all machinery, powered or not , at no RAM cost other than a variable declaration.

**Comment by DGamerL** [comment]
> > ```
> > * Moving it to `/obj/machinery`, which _technically_ is fine but it'd give every piece of machinery a nested list with 25 integers, which can clutter VV a decent amount
> > ```
> 
> Why not just initialize the list only when needed? as in all the big machines just having a set filler proc() in their initialize, this way you avoid both unnecesarry subtypes and provide the functionality for all machinery, powered or not , at no RAM cost other than a variable declaration.

Very good point, I'm gonna do this

**Comment by DGamerL** [comment]
@MLGTASTICa it is done

**Review by Burzah — APPROVED** [review]

File: code/datums/components/multitile.dm [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+			var/obj/structure/filler/new_filler = new(filler_turf)
+			all_fillers += new_filler
+		current_width += 1
+		if(current_width == max_width)
+			current_height += 1
+			current_width = 0
+
+/datum/component/multitile/Destroy(force, silent)
+	QDEL_LIST_CONTENTS(all_fillers)
+	. = ..()
```
Comment (lewcc) [inline code/datums/components/multitile.dm:54]:
```suggestion
	return ..()
```

File: code/datums/components/multitile.dm [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+		machine.deconstruct()
+		return COMPONENT_INCOMPATIBLE
+
+	if(owner.y + distance_from_center > world.maxy || owner.y - distance_from_center < 1)
+		var/obj/machinery/machine = parent
+		machine.deconstruct()
+		return COMPONENT_INCOMPATIBLE
+
+	var/max_height = length(new_filler_map)
+	var/max_width = length(new_filler_map[1]) //it should have the same length on every row
```
Comment (lewcc) [inline code/datums/components/multitile.dm:38]: Please add some sanity checks to ensure that this is indeed the case.

File: code/datums/components/multitile.dm [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+		list(0, 0, 0,		   0, 0),
+		list(0, 0, 0,		   0, 0),
+		list(0, 0, MACH_CENTER, 0, 0),
+		list(0, 0, 0,		   0, 0),
+		list(0, 0, 0,		   0, 0)
+	)
+ */
+
+//distance_from_center does not include src itself
+/datum/component/multitile/Initialize(distance_from_center, new_filler_map)
```
Comment (lewcc) [inline code/datums/components/multitile.dm:19]: I don't really see why we need to specify `distance_from_center` as an argument when we should already have that information in our provided map. 

I think you could probably parse this map a little bit to get some more information on it/to provide some sanity checks. You can make sure all the lists are the same length, that MACH_CENTER is indeed in the center (and somewhere in the lists, for that matter), and then determine the distance from center from there.

**Review by lewcc — CHANGES_REQUESTED** [review]
This will be pretty nice to have, excited to see it.

Got some comments on loading the map in, just to make sure it's as safe as can be. Working with nested lists in byond feels like it's worth double-checking.

File: code/datums/components/multitile.dm (hunk 4 above)
Comment (DGamerL) [reply code/datums/components/multitile.dm:19]: `MACH_CENTER` is just a define for readability, it's just `0`, so no way to really check if it exists. The rest I can do

Comment (lewcc) [reply code/datums/components/multitile.dm:19]: What I mean is that you could make it like -1 or something, and then add some special logic to account for it while processing.

Comment (DGamerL) [reply code/datums/components/multitile.dm:19]: Discussed in discord

**Comment by Daylight2** [comment]
Does this mean that you don't have to interact with the bottom left corner of grav gen anymore? Technology has come so far.

**Comment by DGamerL** [comment]
> Does this mean that you don't have to interact with the bottom left corner of grav gen anymore? Technology has come so far.

Yup! Added the grav gen to this PR!

File: SpacemanDMM.toml [hunk 5]
```diff
@@ -34,3 +34,4 @@ hide_invisible = [
 
 [map_renderer.render_passes]
 smart-cables = false
+gravity-gen = false
```
Comment (lewcc) [inline SpacemanDMM.toml:37]: Might be worth leaving a comment in the file as to why this is the case

**Review by lewcc — APPROVED** [review]
Looks good, thanks for adding the sanity checks. Just one little thing.

File: SpacemanDMM.toml (hunk 5 above)
Comment (DGamerL) [reply SpacemanDMM.toml:38]: Done!

**Pull request merged**